chainer_chemistry/models/gcn.py

```diff
@@ -0,0 +1,132 @@
+"""
+Implementation of Graph Convolutional Network
+(https://arxiv.org/abs/1609.02907)
+"""
+import chainer
+from chainer import cuda, Variable
+from chainer import functions
+import numpy
+
+import chainer_chemistry
+from chainer_chemistry.config import MAX_ATOMIC_NUM
+
+
+class GCNUpdate(chainer.Chain):
+    """GCN sub module for update part
+
+    Args:
+        in_channels (int): input channel dimension
+        out_channels (int): output channel dimension
+    """
+
+    def __init__(self, in_channels, out_channels):
+        super(GCNUpdate, self).__init__()
+        with self.init_scope():
+            self.graph_linear = chainer_chemistry.links.GraphLinear(
+                in_channels, out_channels, nobias=True)
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+    def __call__(self, x, w_adj):
+        # --- Message part ---
+        h = chainer_chemistry.functions.matmul(w_adj, x)
+        # --- Update part ---
+        y = self.graph_linear(h)
+        return y
+
+
+class GCNReadout(chainer.Chain):
+    """GCN sub module for readout part
+
+    Args:
+        in_channels (int): dimension of feature vector associated to each node
+        out_size (int): output dimension of feature vector associated to each graph
+    """
+
+    def __init__(self, in_channels, out_size):
+        super(GCNReadout, self).__init__()
+        with self.init_scope():
+            self.output_weight = chainer_chemistry.links.GraphLinear(
+                in_channels, out_size, nobias=True)
+        self.in_channels = in_channels
+        self.out_size = out_size
+
+    def __call__(self, x):
+        # --- Readout part ---
+        h = self.output_weight(x)
+
+        #
+        h = functions.softmax(h, axis=2)  # softmax along channel axis
+        y = functions.sum(h, axis=1)  # sum along node axis
+        return y
+
```

Comment:
Original paper does not define "Readout" (get feature of "whole graph"), instead, their output is for each node by simply taking softmax only.

So, `output_weight` and `sum` is added by you to extract "graph feature"?
Is my understanding correct?

Reply:
Yes.  To be more precise, I had no idea what to do with readout, so I used the NFP's readout implementation.